### ISSUE TITLE
Need layout system

### ISSUE THREAD

**Issue by danwos** [useblocks/sphinxcontrib-needs#102]
This is a technical spec for a system to allow the user to specify the structure/layout of a need.
It is not based on rst/jinja, as these solutions do not provide all needed functionality, e.g. setting class names.

### needs_layouts
```
needs_layouts = {
  'default' = [
    ('head',  '<<type>>: <<title>>  ID: <<id>>'),
    ('meta', 'status: <<status>>'),
    ('meta',  '**tags**: <<tags>>`),
    ('content', <<content>>')
   ],
   'requirements': [  # focus is on content and status only
    ('content', '**Requirement:**'),
    ('content', '<<content>>'),
    ('content', ' '),  # empty line
    ('meta', 'Requirement status is <<status>>')
    ],
}
```
``needs_layouts`` stores a dictionary, where each element is a layout definition under a given name.
The layout-name can be used to set layout per default for all needs ``needs_layout_default``, to set layout globally  for filter results ``needs_layout_filter`` or to set it directly by using option ``layout`` for needs, needs_extracts or maybe other directives.

A layout_definition is a list of tuples with 2 elements.
Each list-element represents a line in sphinx-documentation and must contain a tuple with following content:

1. section/area name
2. layout string

### section / area name
This name helps us to know, if the line belongs to the title, the meta data area or the content.
We need it to wrap around e.g. meta data area special containers to make it collapsible or to set specific css-classes.

Allowed names are (so far):

* head
* meta
* content

It is allowed to use the same for multiple lines, e.g. ``meta`` for all meta_data lines, even if 2 are coming after ``title`` and 3 are coming after ``content``.

### layout string
A layout string is a single string, which is allowed to use standard rst-syntax and placeholders for special functions, which sphinx-needs provides.

A layout_string like ``<<type>>: <<title>>  ID: <<id>>`` must get separated into the following elements:

1. ``<<type>>``
2.  ``: ``
3.  ``<<title>>``
4. `` ID: ``
5. ``<<id>>``

All elements containing ``<<...>>`` are referencing an existing function, which creates docutiles-nodes, which get embedded at the right place.
The name inside ``<<...>>`` is called layout_function.

Elements, which do not contain ``<<..>>`` are parsed as a rst-string and the resulting docutils-nodes get embedded at the right place.

The the final interpretation of a layout_string is a docutils-node-tree, which gets inject inside a line-node. This line-node is then added to the need-tree.

#### layout_functions
A layout function is referencing an existing python function name.
The return value of this function must be a docutils node(-tree).

A layout_function can have arguments, but no keyword-arguments:
```
<<function(arg1,arg2,arg3)>>
```

So the python representation should look like
```
def function(*args)
```
Arguments must always be from type ``str``.

These arguments can be used as some kind of a configuration. Example:
``<<meta_data(status,tags,my_own_option,links)>>``
This function then should create a node-tree, which contains all the given meta-data values.

Provided functions names so far (not complete):
* type
* title
* id: just id string
* id_link: id string linked to need-ref
* status
* tags
* links: just link-list-string
* links_link: linked links
* content
* meta_data:... (list of option names)

### needs_layout_default
```
needs_layout_default = 'default'
```
Is used, if no filter in ``needs_layout_filter`` has a match.

### needs_layout_filter
```
needs_layout_filter = [
   ('requirements', 'type == "requirement"')
]
```
(layout_name, filter_string)
layout_name must exist in ``needs_layouts`` otherwise it is ignored.

### needs_layout_functions:
Allows the registration or overriding of existing layout functions

```
needs_layout_functions = {
   "id": my_id_function,
   "title": my_title_fcuntion 
}
```

This conf-value must get merged with a default setting, which uses the sphinx-needs internal functions only.
So default values get overridden or if nothing is set for it by user, it is used.

**Comment by danwos** [comment]
**Spec Update**
The above specification is not flexible enough regarding columns.

Therefore sphinx-needs shall provide the following layouts by default:
![image](https://user-images.githubusercontent.com/998700/71341753-08510300-255b-11ea-9732-6522b6d2b612.png)

## needs_layouts
Updated dict-layout:
```
needs_layouts = {
  'default' = {
    'grid': 'default',
    'layout': {
       'head' : [
          '<<type>>: <<title>>  ID: <<id>>',
        ]
        'meta': [
          'status: <<status>>',
          '**tags**: <<tags>>`
        ],
        'content': [
          '<<content>>'
        ]
      }
   },
   'requirements': {  # focus is on content and status only
     'grid': 'content',
     'layout': {
       'content': [
         '**Requirement:**',
         '<<content>>',
         ' ',  # empty line
         'Requirement status is <<status>>'
       ],
     }
  }
}
```

Each own layout must reference a grid-system from above and the used layout-definition must use the names from the grid-elements as the dict-keys.
Keys not used by grid-system will get ignored.
Not all keys used by grid-system need to be used. E.g. ``footer`` does not need to be used, if there is no content planned for it.

Each element in layout must be a list of strings, which will get embedded as line-nodes at the related grid-element-place.

**Comment by danwos** [comment]
Solved with 0.5.0
